Mathlib/LinearAlgebra/Dual.lean

```diff
@@ -589,42 +578,85 @@ theorem erange_coe [FiniteDimensional K V] : LinearMap.range (eval K V) = ⊤ :=
   (Basis.ofVectorSpace K V).eval_range
 #align module.erange_coe Module.erange_coe
 
-variable (K V)
+section IsReflexive
 
-/-- A vector space is linearly equivalent to the dual of its dual space. -/
-def evalEquiv [FiniteDimensional K V] : V ≃ₗ[K] Dual K (Dual K V) :=
-  LinearEquiv.ofBijective
-    (eval K V)-- 60x faster elaboration than using `ker_eq_bot.mp eval_ker` directly:
-    ⟨by
-      rw [← ker_eq_bot]
-      apply eval_ker K V, range_eq_top.mp erange_coe⟩
-#align module.eval_equiv Module.evalEquiv
+open Function
 
-/-- The isomorphism `Module.evalEquiv` induces an order isomorphism on subspaces. -/
-def mapEvalEquiv [FiniteDimensional K V] : Subspace K V ≃o Subspace K (Dual K (Dual K V)) :=
-  Submodule.orderIsoMapComap (evalEquiv K V)
-#align module.map_eval_equiv Module.mapEvalEquiv
+variable (R M : Type _) [CommRing R] [AddCommGroup M] [Module R M]
 
-variable {K V}
+/-- A reflexive module is one for which the natural map to its double dual is a bijection.
 
-@[simp]
-theorem evalEquiv_toLinearMap [FiniteDimensional K V] :
-    (evalEquiv K V).toLinearMap = Dual.eval K V :=
-  rfl
+Any finitely-generated free module (and thus any finite-dimensional vector space) is reflexive.
+See `Module.IsReflexive.of_finite_of_free`. -/
+class IsReflexive : Prop where
+  /-- A reflexive module is one for which the natural map to its double dual is a bijection. -/
+  bijective_dual_eval' : Bijective $ Dual.eval R M
+
+lemma bijective_dual_eval [IsReflexive R M] : Bijective $ Dual.eval R M :=
+  IsReflexive.bijective_dual_eval'
+
+instance IsReflexive.of_finite_of_free [Finite R M] [Free R M] : IsReflexive R M := by
+  cases' h : subsingleton_or_nontrivial R
+  · have := Module.subsingleton R M
+    exact ⟨⟨fun x y _ ↦ by simp, fun x ↦ ⟨0, by simp⟩⟩⟩
```

Comment:
This could be a standalone `IsReflective.of_subsingleton [Subsingleton M]`, right?

Reply:
I actually considered this but I didn't bother because it's just a special case of this lemma, specifically:
```
example [Subsingleton M] : Finite R M := inferInstance
example [Subsingleton M] : Free R M := inferInstance
```